MatrixMultiplication/MatrixMultiplication/Matrix.cs

```diff
@@ -0,0 +1,196 @@
+/// class that implements a matrix data structure
+/// </summary>
+public class Matrix
+{
+    private readonly int[,] data;
+
+    public int Rows => data.GetLength(0);
+
+    public int Columns => data.GetLength(1);
+
+    private static Random rand = new();
+
+    public Matrix(string path)
+    {
+        string matrixDimensionsPattern = @"^\d+ \d+$";
+        using var reader = new StreamReader(path);
+        var firstLine = reader.ReadLine();
+        if (firstLine == null)
+            throw new FormatException("file is empty");
+        if (!Regex.IsMatch(firstLine, matrixDimensionsPattern))
+            throw new FormatException("first line must contain two positive integers separated by space");
+        string[] dimensions = firstLine.Split();
+        int rows = int.Parse(dimensions[0]);
+        int columns = int.Parse(dimensions[1]);
+        data = new int[rows, columns];
+        string rowsPattern = @"^(\d+ ?)*$";
+        for (int i = 0; i < rows; i++)
+        {
+            var line = reader.ReadLine();
+            if (line == null)
+                throw new FormatException("actual number of rows is not equal to stated");
+            if (!Regex.IsMatch(line, rowsPattern))
+                throw new FormatException("rows must contain only numbers");
+            string[] intsInStrings = line.Split();
+            if (intsInStrings.Length != columns)
+                throw new FormatException($"length of {i}th row ({intsInStrings.Length}) is not equal to stated number of columns ({columns})");
+            for (int j = 0; j < intsInStrings.Length; j++)
+            {
+                data[i, j] = int.Parse(intsInStrings[j]);
+            }
+        }
+    }
+
+    public Matrix(int[,] data)
+    {
+        int rows = data.GetLength(0);
+        int columns = data.GetLength(1);
+        this.data = new int[rows, columns];
+        for (int i = 0; i < rows; i++)
+        {
+            for (int j = 0; j < columns; j++)
+            {
+                this.data[i, j] = data[i, j];
+            }
+        }
+    }
+
+    /// <summary>
+    /// multiplies matrix by another matrix without additional threads
+    /// </summary>
+    /// <param name="other">matrix to multiply by</param>
+    /// <returns>result of multiplication</returns>
+    /// <exception cref="InvalidOperationException">thrown if multiplication is not defined for given matrices</exception>
+    public Matrix MultiplyBy(Matrix other)
+    {
+        if (this.Columns != other.Rows)
+            throw new InvalidOperationException("matrices' dimensions are not compliant for multiplication");
+        int[,] result = new int[this.Rows, other.Columns];
+        for (int i = 0; i < this.Rows; i++)
+        {
+            for (int j = 0; j < other.Columns; j++)
+            {
+                for (int k = 0; k < this.Columns; k++)
+                {
+                    result[i, j] += this.data[i, k] * other.data[k, j];
+                }
+            }
```

Comment:
можно написать короче:
```suggestion
            for (int j = 0; j < other.Columns; j++)
            {
                    result[i, j] = Enumerable.Range(0, this.Columns).Select(k => this.data[i, k] * other.data[k, j]);
            }
```